File: sections/14_core_patterns_mixed_contest/problems/r_coverage_patches/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n;
    long long target;
    cin >> n >> target;
    vector<long long> values(n);
    for (long long &value : values) cin >> value;
    sort(values.begin(), values.end());
    long long covered = 0;
    int index = 0, patches = 0;
    while (covered < target) {
        if (index < n && values[index] <= covered + 1) {
            covered += values[index++];
        } else {
            covered += covered + 1;
            ++patches;
        }
    }
    cout << patches << '\n';
}

File: sections/14_core_patterns_mixed_contest/problems/r_coverage_patches/solution.py
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    target = data[1]
    values = sorted(data[2:])
    covered = index = patches = 0
    while covered < target:
        if index < len(values) and values[index] <= covered + 1:
            covered += values[index]
            index += 1
        else:
            covered += covered + 1
            patches += 1
    print(patches)


if __name__ == "__main__":
    main()
